File: examples/sidecar-js/decision-audit-reporter/test/event-tracker.test.ts
import { describe, expect, it } from 'vitest';

import { EventTracker } from '../src/event-tracker.js';

function buildDmnFinishedEvent(overrides: Record<string, unknown> = {}): Record<string, unknown> {
  return {
    flowNodeInstanceId: 'fni-1',
    processInstanceId: 'pi-1',
    flowNodeId: 'BRT_1',
    flowNodeType: 'business_rule_task',
    terminalState: 'completed',
    occurredAt: '2026-05-20T12:00:00Z',
    payload: {
      typeProperties: {
        mode: 'dmn',
        decision_ref: 'employee-benefits',
        duration_us: 1500,
        matched_rules: ['rule_1'],
      },
    },
    ...overrides,
  };
}

describe('EventTracker', () => {
  it('records FNI finished events for BRT+DMN only', () => {
    const tracker = new EventTracker();
    tracker.handleEvent(buildDmnFinishedEvent());

    expect(tracker.getCount()).toBe(1);
    expect(tracker.getTrackedFnis()).toEqual(['fni-1']);
    expect(tracker.getEvents()[0].flowNodeInstanceId).toBe('fni-1');
  });

  it('ignores non-BRT and non-DMN events', () => {
    const tracker = new EventTracker();

    tracker.handleEvent({
      ...buildDmnFinishedEvent(),
      flowNodeType: 'service_task',
    });

    tracker.handleEvent({
      ...buildDmnFinishedEvent(),
      payload: { typeProperties: { mode: 'feel' } },
    });

    tracker.handleEvent({
      flowNodeType: 'business_rule_task',
      flowNodeInstanceId: 'fni-2',
      payload: { typeProperties: { mode: 'dmn' } },
    });

    expect(tracker.getCount()).toBe(1);
    expect(tracker.getTrackedFnis()).toEqual(['fni-2']);
  });

  it('getCount and getTrackedFnis work correctly', () => {
    const tracker = new EventTracker();
    tracker.handleEvent(buildDmnFinishedEvent({ flowNodeInstanceId: 'fni-a' }));
    tracker.handleEvent(buildDmnFinishedEvent({ flowNodeInstanceId: 'fni-b' }));

    expect(tracker.getCount()).toBe(2);
    expect(tracker.getTrackedFnis()).toEqual(['fni-a', 'fni-b']);
  });

  it('reset clears state', () => {
    const tracker = new EventTracker();
    tracker.handleEvent(buildDmnFinishedEvent());
    tracker.reset();

    expect(tracker.getCount()).toBe(0);
    expect(tracker.getTrackedFnis()).toEqual([]);
    expect(tracker.getEvents()).toEqual([]);
  });
});

File: examples/plugins/business_rules/decision_audit_reporter/test/event_tracker_test.exs
defmodule Examples.BusinessRules.DecisionAuditReporter.EventTrackerTest do
  use ExUnit.Case, async: true

  alias Examples.BusinessRules.DecisionAuditReporter.EventTracker

  setup do
    tracker_name = :"event_tracker_test_#{:erlang.unique_integer([:positive])}"
    {:ok, _pid} = EventTracker.start_link(name: tracker_name)
    {:ok, tracker_name: tracker_name}
  end

  defp tracked_event(flow_node_instance_id) do
    %{
      flow_node_instance_id: flow_node_instance_id,
      process_instance_id: "process-instance-1",
      flow_node_id: "BRT_1"
    }
  end

  test "records FNI finished events", %{tracker_name: tracker_name} do
    :ok = EventTracker.track(tracked_event("fni-1"), name: tracker_name)

    assert EventTracker.get_count(name: tracker_name) == 1
    assert EventTracker.get_tracked_flow_node_instance_ids(name: tracker_name) == ["fni-1"]
    assert hd(EventTracker.get_events(name: tracker_name)).flow_node_instance_id == "fni-1"
  end

  test "get_count and get_tracked_flow_node_instance_ids work correctly", %{
    tracker_name: tracker_name
  } do
    :ok = EventTracker.track(tracked_event("fni-a"), name: tracker_name)
    :ok = EventTracker.track(tracked_event("fni-b"), name: tracker_name)

    assert EventTracker.get_count(name: tracker_name) == 2
    assert EventTracker.get_tracked_flow_node_instance_ids(name: tracker_name) == ["fni-a", "fni-b"]
  end

  test "reset clears state", %{tracker_name: tracker_name} do
    :ok = EventTracker.track(tracked_event("fni-1"), name: tracker_name)
    :ok = EventTracker.reset(name: tracker_name)

    assert EventTracker.get_count(name: tracker_name) == 0
    assert EventTracker.get_tracked_flow_node_instance_ids(name: tracker_name) == []
    assert EventTracker.get_events(name: tracker_name) == []
  end
end
